library/core/src/iter/adapters/array_chunks.rs

```diff
@@ -0,0 +1,146 @@
+use core::iter::FusedIterator;
+use core::mem::{self, MaybeUninit};
+use core::ptr;
+
+/// An iterator that yields the elements of another iterator in
+/// chunks of size `N`.
+///
+/// This `struct` is created by the [`array_chunks`] method on [`Iterator`]. See
+/// its documentation for more.
+///
+/// [`array_chunks`]: Iterator::array_chunks
+#[unstable(feature = "iter_array_chunks", issue = "none")]
+#[derive(Debug)]
+pub struct ArrayChunks<I: Iterator, const N: usize> {
+    iter: I,
+    buffer: [MaybeUninit<I::Item>; N],
+    init: usize,
+}
+
+#[unstable(feature = "iter_array_chunks", issue = "none")]
+impl<I: Iterator, const N: usize> Drop for ArrayChunks<I, N> {
+    fn drop(&mut self) {
+        // SAFETY: This is safe: `remainder_mut` returns exactly the sub-slice
+        // of elements that were initialized but not yielded and so have yet
+        // to be dropped.
+        unsafe {
+            ptr::drop_in_place(self.remainder_mut());
+        }
+    }
+}
+
+impl<I: Iterator, const N: usize> ArrayChunks<I, N> {
+    pub(in crate::iter) fn new(iter: I) -> Self {
+        Self { iter, init: 0, buffer: MaybeUninit::uninit_array() }
+    }
+
+    /// Returns the remainder of the elements yielded by the original
+    /// iterator that were insufficient to fill another chunk. The
+    /// returned slice has at most `N-1` elements.
+    #[unstable(feature = "iter_array_chunks", issue = "none")]
+    pub fn remainder(&self) -> &[I::Item] {
+        // SAFETY: We know that all elements before `init` are properly initialized.
+        unsafe { MaybeUninit::slice_assume_init_ref(&self.buffer[..self.init]) }
+    }
+
+    /// Returns the remainder of the elements yielded by the original
+    /// iterator that were insufficient to fill another chunk. The
+    /// returned slice has at most `N-1` elements.
+    #[unstable(feature = "iter_array_chunks", issue = "none")]
+    pub fn remainder_mut(&mut self) -> &mut [I::Item] {
+        // SAFETY: We know that all elements before `init` are properly initialized.
+        unsafe { MaybeUninit::slice_assume_init_mut(&mut self.buffer[..self.init]) }
+    }
+}
+
+#[unstable(feature = "iter_array_chunks", issue = "none")]
+impl<I: Iterator, const N: usize> Iterator for ArrayChunks<I, N> {
+    type Item = [I::Item; N];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.init < N {
+            self.buffer[self.init] = MaybeUninit::new(self.iter.next()?);
+            self.init += 1;
+        }
+        self.init = 0;
+        // SAFETY: This is safe: `MaybeUninit<T>` is guaranteed to have the same layout
+        // as `T` and the entire array has just been initialized.
+        unsafe { Some(mem::transmute_copy(&self.buffer)) }
```

Comment:
Reusing memory (going through `self`) and then transmute-copying it to return it seems like it would optimize badly. Benchmarks exercising this adapter together with a for-in loop as baseline would help.

Reply:
I have added some benchmarks comparing it to looping over a mapped iterator producing arrays and the `ArrayChunks` iterator for slices.
```
test iter::bench_array_map                                      ... bench:     437,783 ns/iter (+/- 13,664)
test iter::bench_iter_array_chunks                              ... bench:   2,684,281 ns/iter (+/- 206,116)
test iter::bench_iter_array_chunks_fold                         ... bench:   2,217,095 ns/iter (+/- 127,330)
test iter::bench_slice_array_chunks                             ... bench:     335,405 ns/iter (+/- 99,203)
```

Reply:
Is there a better way to run specific benchmarks for `core`, than running all of them with `./x.py bench library/core`? It doesn't seem to accept that standard filter that cargo bench uses.

Reply:
I managed to improve the performance by quite a bit. Now it only takes twice as long as passing an array through an iterator pipeline. I also removed the `transmute_copy` (slight increase in performance).

```
test iter::bench_array_map                                      ... bench:     412,536 ns/iter (+/- 19,784)
test iter::bench_iter_array_chunks_fold                         ... bench:     856,663 ns/iter (+/- 67,238)
test iter::bench_iter_array_chunks_loop                         ... bench:   1,678,215 ns/iter (+/- 78,573)
test iter::bench_iter_array_chunks_ref_fold                     ... bench:   1,511,018 ns/iter (+/- 86,835)
test iter::bench_slice_array_chunks                             ... bench:     337,726 ns/iter (+/- 83,062)
```